--- masterfiles/update.cf.orig	2013-02-08 13:21:26.000000000 +0300
+++ masterfiles/update.cf	2013-02-08 13:23:02.000000000 +0300
@@ -48,7 +48,6 @@
                       "cf-agent",
                       "cf-execd",
                       "cf-key",
-                      "cf-know",
                       "cf-monitord",
                       "cf-promises",
                       "cf-report",
@@ -166,13 +165,13 @@
       depth_search => u_recurse_basedir("inf"),
             action => u_immediate;
 
-   "/usr/local/sbin/$(agents)"
+   "$(sys.workdir)/bin/$(agents)"
                 comment => "Create symlinks of CFE binaries in /usr/local/sbin",
                  handle => "update_files_sbin_$(agents)",
       move_obstructions => "true",
-              link_from => u_ln_s("$(sys.workdir)/bin/$(agents)");
+              link_from => u_ln_s("/usr/sbin/$(agents)");
 
-   "/usr/local/sbin/$(agents).cfsaved"
+   "$(sys.workdir)/bin/$(agents).cfsaved"
       comment => "Remove all .cfsaved file extension",
        handle => "update_files_remove_$(agent)_cfsaved",
        delete => u_tidy;
