diff -Naur a/freeradius-server-3.0.10/src/modules/rlm_sql/drivers/rlm_sql_postgresql/rlm_sql_postgresql.c b/freeradius-server-3.0.10/src/modules/rlm_sql/drivers/rlm_sql_postgresql/rlm_sql_postgresql.c
--- a/freeradius-server-3.0.10/src/modules/rlm_sql/drivers/rlm_sql_postgresql/rlm_sql_postgresql.c	2015-10-05 21:22:28.000000000 +0300
+++ b/freeradius-server-3.0.10/src/modules/rlm_sql/drivers/rlm_sql_postgresql/rlm_sql_postgresql.c	2016-01-21 12:43:37.000000000 +0300
@@ -140,6 +140,10 @@
 			db_string = talloc_asprintf_append(db_string, " password='%s'", config->sql_password);
 		}
 
+		if (config->connect_timeout_s) {
+			db_string = talloc_asprintf_append(db_string, " connect_timeout=%i", config->connect_timeout_s);
+		}
+
 		if (driver->send_application_name) {
 			db_string = talloc_asprintf_append(db_string, " application_name='%s'", application_name);
 		}
@@ -168,6 +172,10 @@
 			db_string = talloc_asprintf_append(db_string, " password='%s'", config->sql_password);
 		}
 
+		if (config->connect_timeout_s && !strstr(db_string, "connect_timeout=")) {
+			db_string = talloc_asprintf_append(db_string, " connect_timeout=%i", config->connect_timeout_s);
+		}
+
 		if (driver->send_application_name && !strstr(db_string, "application_name=")) {
 			db_string = talloc_asprintf_append(db_string, " application_name='%s'", application_name);
 		}
diff -Naur a/freeradius-server-3.0.10/src/modules/rlm_sql/rlm_sql.c b/freeradius-server-3.0.10/src/modules/rlm_sql/rlm_sql.c
--- a/freeradius-server-3.0.10/src/modules/rlm_sql/rlm_sql.c	2015-10-05 21:22:28.000000000 +0300
+++ b/freeradius-server-3.0.10/src/modules/rlm_sql/rlm_sql.c	2016-01-21 12:41:37.000000000 +0300
@@ -84,6 +84,7 @@
 	{ "port", FR_CONF_OFFSET(PW_TYPE_INTEGER, rlm_sql_config_t, sql_port), "0" },
 	{ "login", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_sql_config_t, sql_login), "" },
 	{ "password", FR_CONF_OFFSET(PW_TYPE_STRING | PW_TYPE_SECRET, rlm_sql_config_t, sql_password), "" },
+	{ "connect_timeout", FR_CONF_OFFSET(PW_TYPE_INTEGER, rlm_sql_config_t, connect_timeout_s), "0" },
 	{ "radius_db", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_sql_config_t, sql_db), "radius" },
 	{ "read_groups", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_sql_config_t, read_groups), "yes" },
 	{ "read_profiles", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_sql_config_t, read_profiles), "yes" },
