--- Radius.pm	2010-01-14 10:20:50.000000000 +0200
+++ Radius.pm.patched	2010-02-02 13:39:06.000000000 +0200
@@ -31,8 +31,8 @@
 @ISA = qw(Exporter);
 @EXPORT = qw(ACCESS_REQUEST ACCESS_ACCEPT ACCESS_REJECT
 			ACCOUNTING_REQUEST ACCOUNTING_RESPONSE ACCOUNTING_STATUS
-			DISCONNECT_REQUEST
-			COA_REQUEST);
+			DISCONNECT_REQUEST DISCONNECT_ACCEPT DISCONNECT_REJECT
+			COA_REQUEST COA_ACCEPT COA_REJECT);
 $VERSION = '0.17';
 
 my (%dict_id, %dict_name, %dict_val, %dict_vendor_id, %dict_vendor_name );
@@ -56,7 +56,11 @@
 use constant ACCOUNTING_RESPONSE          => 5;
 use constant ACCOUNTING_STATUS            => 6;
 use constant DISCONNECT_REQUEST           => 40;
+use constant DISCONNECT_ACCEPT		  => 41;
+use constant DISCONNECT_REJECT		  => 42;
 use constant COA_REQUEST                  => 43; 
+use constant COA_ACCEPT                   => 44;
+use constant COA_REJECT                   => 45;
 
 my $HMAC_MD5_BLCKSZ = 64;
 my $RFC3579_MSG_AUTH_ATTR_ID = 80;
