--- build-aux/ltmain.sh.orig	2014-04-16 09:30:26.000000000 +0300
+++ build-aux/ltmain.sh	2014-04-16 09:31:21.000000000 +0300
@@ -5664,6 +5664,15 @@
 	    *" $arg "*) ;;
 	    * ) func_append new_inherited_linker_flags " $arg" ;;
 	esac
+      # As we are forced to pass -nostdlib to g++ during linking, the option
+      # -pthread{,s} is not in effect;  add the -lpthread to $deplist
+      # explicitly to link correctly.
+      if test "$tagname" = CXX -a x"$with_gcc" = xyes; then
+        case "$arg" in
+          -pthread*) func_append deplibs " -lpthread" ;;
+        esac
+      fi
+	
 	continue
 	;;
 
