--- flow-tools-0.68/lib/ftlib.h.orig	2005-05-10 18:51:33.000000000 +0300
+++ flow-tools-0.68/lib/ftlib.h	2009-09-21 17:36:37.000000000 +0300
@@ -416,8 +416,8 @@
 };
 
 struct fttime {
-  u_int32 secs;
-  u_int32 msecs;
+  time_t secs;
+  time_t msecs;
 };
 
 struct ftver {
@@ -2710,7 +2710,7 @@
 int ftfile_dump(struct ftfile_entries *fte);
 struct ftfile_entry *ftfile_entry_new(int len);
 void ftfile_entry_free(struct ftfile_entry *entry);
-int ftfile_mkpath(u_int32 ftime, int nest);
+int ftfile_mkpath(time_t ftime, int nest);
 void ftfile_pathname(char *buf, int bsize, int nest, struct ftver ftv,
  int done, u_int32 ftime);
