--- plugins/forbid/cforbidden.cpp.orig	2011-07-25 14:52:25.000000000 +0300
+++ plugins/forbid/cforbidden.cpp	2011-07-25 14:52:51.000000000 +0300
@@ -106,11 +106,11 @@
 
 void cForbidden::AddFields()
 {
-	AddCol("word","varchar(100)","", false,mModel.mWord);
+	AddCol("word","varchar(500)","", false,mModel.mWord);
 	AddPrimaryKey("word");
 	AddCol("check_mask","tinyint(4)","1",true,mModel.mCheckMask); // mask 1 -> public, mask 2 -> private
 	AddCol("afclass","tinyint(4)", "4", true, mModel.mAfClass); // affected class. normal=1, vip=2, cheef=3, op=4, admin=5, master=10
-	AddCol("banreason", "varchar(50)", "", true, mModel.mReason);// reason of kick
+	AddCol("banreason", "varchar(100)", "", true, mModel.mReason);// reason of kick
 	mMySQLTable.mExtra =" PRIMARY KEY(word)";
 }
 