--- src/crypto-dmc.c.orig	2011-07-25 12:00:23.000000000 +0300
+++ src/crypto-dmc.c	2011-07-25 12:01:08.000000000 +0300
@@ -29,7 +29,7 @@
 	struct crypt_device *cd;
 	const char *device = path;
 	char *loop_device;
-	int ret;
+	int ret,ret2;
 
 	if (!blkdev) {
 		ret = pmt_loop_setup(path, &loop_device, true);
@@ -53,9 +53,13 @@
 			ret = true;
 		/* else keep ret as-is */
 		crypt_free(cd);
-	}
-	if (!blkdev)
-		pmt_loop_release(loop_device);
+ 	}
+        if (!blkdev) {
+               ret2 = pmt_loop_release(loop_device);
+               if (ret2 < 0)
+                       fprintf(stderr, "pmt_loop_release: %s\n",
+                               strerror(-ret));
+        }
 	return ret;
 }
 
