--- Radius.pm	2009-09-02 14:27:29.000000000 +0300
+++ Radius.pm.patched	2010-01-11 11:08:54.000000000 +0200
@@ -31,7 +31,7 @@
 @ISA = qw(Exporter);
 @EXPORT = qw(ACCESS_REQUEST ACCESS_ACCEPT ACCESS_REJECT
 			 ACCOUNTING_REQUEST ACCOUNTING_RESPONSE ACCOUNTING_STATUS
-			DISCONNECT_REQUEST);
+			DISCONNECT_REQUEST COA_REQUEST);
 $VERSION = '0.15';
 
 my (%dict_id, %dict_name, %dict_val, %dict_vendor_id, %dict_vendor_name );
@@ -55,6 +55,7 @@
 use constant ACCOUNTING_RESPONSE          => 5;
 use constant ACCOUNTING_STATUS            => 6;
 use constant DISCONNECT_REQUEST           => 40;
+use constant COA_REQUEST		  => 43;
 
 sub new {
 	my $class = shift;
@@ -107,7 +108,7 @@
 		$request_id = ($request_id + 1) & 0xff;
 	}
 	$self->set_error;    
-	if ($type == ACCOUNTING_REQUEST || $type == DISCONNECT_REQUEST) {
+	if ($type == ACCOUNTING_REQUEST || $type == DISCONNECT_REQUEST || $type == COA_REQUEST) {
 	  $self->{'authenticator'} = "\0" x 16;
 	  $self->{'authenticator'} =
 	    $self->calc_authenticator($type, $request_id, $length)
