--- src/cluster.c.orig	2016-08-29 16:12:46.000000000 +0300
+++ src/cluster.c	2016-08-29 16:13:12.000000000 +0300
@@ -937,7 +937,11 @@
 	struct AANode *node;
 	const char *username;
 
-	username = GetUserNameFromId(user_oid);
+	username = GetUserNameFromId(user_oid
+#if PG_VERSION_NUM >= 90500
+                                               ,false
+#endif
+       	);
 
 	node = aatree_search(&cluster->userinfo_tree, (uintptr_t)username);
 	if (node) {
