--- proto/ospf/packet.orig	2010-03-24 21:43:58.000000000 +0200
+++ proto/ospf/packet.c	2010-03-24 21:44:23.000000000 +0200
@@ -179,7 +179,7 @@
         return 0;
       }
 
-      if (ntohs(pkt->length) + OSPF_AUTH_CRYPT_SIZE != size)
+      if (ntohs(pkt->length) + OSPF_AUTH_CRYPT_SIZE > size)
       {
         OSPF_TRACE(D_PACKETS, "OSPF_auth: size mismatch (%d vs %d)",
 	  ntohs(pkt->length) + OSPF_AUTH_CRYPT_SIZE, size);
